command/agent/exec/exec.go

```diff
@@ -0,0 +1,288 @@
+package exec
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/consul-template/child"
+	ctconfig "github.com/hashicorp/consul-template/config"
+	"github.com/hashicorp/consul-template/manager"
+	"github.com/hashicorp/go-hclog"
+	"go.uber.org/atomic"
+
+	"github.com/hashicorp/vault/command/agent/config"
+	"github.com/hashicorp/vault/command/agent/internal/ctmanager"
+	"github.com/hashicorp/vault/helper/useragent"
+	"github.com/hashicorp/vault/sdk/helper/pointerutil"
+)
+
+type ServerConfig struct {
+	Logger      hclog.Logger
+	AgentConfig *config.Config
+
+	Namespace string
+
+	// LogLevel is needed to set the internal Consul Template Runner's log level
+	// to match the log level of Vault Agent. The internal Runner creates it's own
+	// logger and can't be set externally or copied from the Template Server.
+	//
+	// LogWriter is needed to initialize Consul Template's internal logger to use
+	// the same io.Writer that Vault Agent itself is using.
+	LogLevel  hclog.Level
+	LogWriter io.Writer
+}
+
+type Server struct {
+	// config holds the ServerConfig used to create it. It's passed along in other
+	// methods
+	config *ServerConfig
+
+	// runner is the consul-template runner
+	runner *manager.Runner
+
+	// lookupMap is a list of templates indexed by their consul-template ID. This
+	// is used to ensure all Vault templates have been rendered before returning
+	// from the runner in the event we're using exit after auth.
+	lookupMap map[string][]*ctconfig.TemplateConfig
+
+	stopped *atomic.Bool
+
+	logger hclog.Logger
+
+	proc        *child.Child
+	procStarted bool
+	procLock    sync.Mutex
+
+	// exit channel of the child process
+	exitCh <-chan int
+}
+
+type ProcessExitError struct {
+	ExitCode int
+}
+
+func (e *ProcessExitError) Error() string {
+	return fmt.Sprintf("process exited with %d", e.ExitCode)
+}
+
+func NewServer(cfg *ServerConfig) *Server {
+	server := Server{
+		stopped:     atomic.NewBool(false),
+		logger:      cfg.Logger,
+		config:      cfg,
+		procStarted: false,
+		// exitCh: make(<-chan int),
+	}
+
+	return &server
+}
+
+func (s *Server) Run(ctx context.Context, incomingVaultToken chan string) error {
+	latestToken := new(string)
+	s.logger.Info("starting exec server")
+	defer func() {
+		s.logger.Info("exec server stopped")
+	}()
+
+	if len(s.config.AgentConfig.EnvTemplates) == 0 || s.config.AgentConfig.Exec == nil {
+		s.logger.Info("no env templates or exec config, exiting")
+		<-ctx.Done()
```

Comment:
Will this block on context.Done() channel?

Reply:
It certainly will it's just a matter of if that is intended.

Reply:
Is this meant to instead signal that a context should be terminated? In which case you'd need a canclable context. More reading https://stackoverflow.com/questions/52799280/context-confusion-regarding-cancellation

Reply:
not sure I understand the question here.
this is to handle the context coming in from the command
https://github.com/hashicorp/vault/blob/35e2c1665f009183088387532e17d02ded312e18/command/agent.go#L280-L284

Reply:
This is not a typical usage of `ctx.Done()`. In this context, we are checking if the config is invalid and exiting early. There is no point waiting for the context to be cancelled externally before exiting early (unless I'm missing something).

Side note: this whole if-check could probably be skipped since we already validate that both env_template and exec are specified together during the validation.

Reply:
TODO: return nil instead of blocking